controllers/webhook.go

```diff
@@ -0,0 +1,157 @@
+	return nil
+}
+
+func (r *DSPAReconciler) CleanUpWebhookIfUnused(ctx context.Context, dspa *dspav1.DataSciencePipelinesApplication, params *DSPAParams) error {
+	log := r.Log.WithValues("namespace", params.DSPONamespace, "dspa_name", dspa.Name)
+
+	hasK8sDSPAs, err := r.checkAvailableKubernetesDSPAs(ctx, dspa.Name, dspa.Namespace)
+	if err != nil {
+		log.Error(err, "Failed to check for other DSPAs with 'kubernetes' storage")
+		return err
+	}
+
+	if !hasK8sDSPAs {
+		log.Info("No other DSPAs with PipelineStorage 'kubernetes' found. Cleaning up webhook resources.")
+		if err := r.cleanupWebhookResources(ctx, params.DSPONamespace); err != nil {
+			log.Error(err, "Failed to clean up webhook resources")
+			return err
+		}
+	}
+
+	log.Info("Webhook resources cleanup complete.")
+	return nil
+}
+
+func (r *DSPAReconciler) cleanupWebhookResources(ctx context.Context, namespace string) error {
+	log := r.Log.WithValues("namespace", namespace)
+	webhookConfigName := "pipelineversions.pipelines.kubeflow.org"
+	webhookNamespaced := types.NamespacedName{Name: k8sWebhookName, Namespace: namespace}
+
+	// Delete MutatingWebhookConfiguration
+	mutating := &admv1.MutatingWebhookConfiguration{}
+	if err := r.Get(ctx, types.NamespacedName{Name: webhookConfigName}, mutating); err == nil {
+		log.Info("Deleting MutatingWebhookConfiguration", "webhook", mutating)
+		if err := r.Delete(ctx, mutating); err != nil {
+			log.Error(err, "Failed to delete MutatingWebhookConfiguration")
+			return err
+		}
+	}
+
+	// Delete ValidatingWebhookConfiguration
+	validating := &admv1.ValidatingWebhookConfiguration{}
+	if err := r.Get(ctx, types.NamespacedName{Name: webhookConfigName}, validating); err == nil {
+		log.Info("Deleting ValidatingWebhookConfiguration", "webhook", validating)
+		if err := r.Delete(ctx, validating); err != nil {
+			log.Error(err, "Failed to delete ValidatingWebhookConfiguration")
+			return err
+		}
+	}
+
+	// Delete Role
+	role := &rbacv1.ClusterRole{}
+	if err := r.Get(ctx, types.NamespacedName{Name: k8sWebhookName}, role); err == nil {
+		log.Info("Deleting webhook Role", "role", role)
+		if err := r.Delete(ctx, role); err != nil {
+			log.Error(err, "Failed to delete webhook Role")
+			return err
+		}
+	}
+
+	// Delete RoleBinding
+	roleBinding := &rbacv1.ClusterRoleBinding{}
+	if err := r.Get(ctx, types.NamespacedName{Name: k8sWebhookName}, roleBinding); err == nil {
+		log.Info("Deleting webhook RoleBinding", "roleBinding", roleBinding)
+		if err := r.Delete(ctx, roleBinding); err != nil {
+			log.Error(err, "Failed to delete webhook RoleBinding")
+			return err
+		}
+	}
+
+	// Delete Service
+	svc := &corev1.Service{}
+	if err := r.Get(ctx, webhookNamespaced, svc); err == nil {
+		log.Info("Deleting webhook Service", "service", svc.Name)
+		if err := r.Delete(ctx, svc); err != nil {
+			log.Error(err, "Failed to delete webhook Service")
+			return err
+		}
+	}
+
+	// Delete deployments
+	deploy := &appsv1.Deployment{}
```

Comment:
Could you please delete the deployment just after the webhooks? Otherwise it'll be running a service account with no permissions if the service fails or role binding fails to be deleted.